src/toil/lib/toilDocker.py

```diff
@@ -0,0 +1,213 @@
+        command = [' '.join(p) for p in [map(pipes.quote, parameters)]]
+        _logger.info("Calling docker with: " + repr(command))
+
+    # If the 'parameters' lists are empty, they are respecified as None, which tells the API
+    # to simply create and run the container
+    else:
+        entrypoint = None
+        command = None
+
+    workDir = os.path.abspath(workDir)
+
+    # Ensure the user has passed a valid value for deferParam
+    assert deferParam in (None, FORGO, STOP, RM), 'Please provide a valid value for deferParam.'
+
+    client = docker.from_env()
+
+    if deferParam == STOP:
+        job.defer(_dockerStop, containerName, client)
+
+    if deferParam == FORGO:
+        removeUponExit = False
+    elif deferParam == RM:
+        removeUponExit = True
+        job.defer(_dockerKill, containerName, client)
+    # really shaky on this ->>
+    elif removeUponExit is True:
+        job.defer(_dockerKill, containerName, client)
+
+    try:
+        out = client.containers.run(image=tool,
+                                    command=command,
+                                    working_dir=workDir,
+                                    entrypoint=entrypoint,
+                                    name=containerName,
+                                    detach=detach,
+                                    volumes={workDir: {'bind': data_dir, 'mode': 'rw'}},
+                                    auto_remove=removeUponExit,
+                                    remove=removeUponExit,
+                                    log_config=logDriver,
+                                    user=str(thisUser) + ":" + str(thisGroup))
+        return out
+    # If the container exits with a non-zero exit code and detach is False.
+    except ContainerError:
+        _logger.error("Docker had non-zero exit.  Check your command: " + repr(command))
+        raise ContainerError(container=containerName, exit_status=1, command=command, image=tool, stderr=stderr)
+    except ImageNotFound:
+        _logger.error("Docker image not found.")
+        raise ImageNotFound("Docker image not found.")
+    except APIError:
+        _logger.error("The server returned an error.")
+        raise APIError("The server returned an error.")
+    except:
+        raise ("An unexpected error occurred while attempting to call docker.")
+
+def _dockerKill(container_name, client):
+    """
+    Immediately kills a container.  Equivalent to "docker kill":
+    https://docs.docker.com/engine/reference/commandline/kill/
+
+    :param container_name: Name of the container being killed.
+    :param client: The docker API client object to call.
+    """
+    try:
+        this_container = client.containers.get(container_name)
+        if this_container.status == 'running':
+            this_container.kill(signal=None)
+            while (this_container.status == 'running'):
+                this_container = client.containers.get(container_name)
```

Comment:
There should maybe be a sleep in this loop somewhere.

Reply:
Updated.